New Greenshot file format

Christian-Schulz opened the pull request:
First of all, I must apologize. The pull request is very large, but there was no other way.

Everything else that is not directly related to the Greenshot file format i moved to other PR's.

https://github.com/greenshot/greenshot/pull/593 (✅merged in main)
https://github.com/greenshot/greenshot/pull/594 (✅merged in main)
https://github.com/greenshot/greenshot/pull/595 (✅merged in main)

---
### Overview

The main reason for this PR is that we need to move away from using BinaryFormatter and separate the serialization functions from the domain classes. 

* This PR implements a new file format (V2) for Greenshot editor files (.greenshot) and templates (.gst), replacing the BinaryFormatter with MessagePack serialization. 
* The implementation is backward compatible for all files up to Greenshot version 1.2.
* I added a new test project with xUnit.
* I refactored most of the .greenshot file related stuff and moved it to the new namespace `Greenshot.Editor.FileFormat`.
* I refactored all `DrawableContainer` classes and removed all [Serializable] attributes.

### Serialization

I chose MessagePack as the new serializer because it is recommended as a replacement. It supports everything we need so far and also allows us to handle future file format changes and content extensions.
There are Greenshot Editor files (.greenshot) and templates (.gst). These are very similar to each other. Everything I write about .greenshot below also applies to .gst.

The `Greenshot.Editor.FileFormat.GreenshotFile` represent a .greenshot file with domain data.
In `Greenshot.Editor.FileFormat.DTO` are Data Transfer Objects that represent a .greenshot file in a lightweight structure. Main class is `GreenshotFileDTO`. 
These DTO classes are serialized using MessagePack.
`Greenshot.Editor.FileFormat.Dto.ConvertDomainToDto.ToDto()` is used to convert `GreenshotFile`  into DTO.
`Greenshot.Editor.FileFormat.Dto.ConvertDtoToDomain.ToDomain()` performs the same operation in reverse.

These changes together ensure that the serialization logic is largely separated from the domain classes.

I have implemented a new versioning scheme for .greenshot files. Previously, the file version was closely linked to the application version. From now on, the versions are separate.
The old files for Greenshot 1.2 and 1.3 are now handled under the new version V1 in `Greenshot.Editor.FileFormat.V1.GreenshotFileV1`.
The new files in `Greenshot.Editor.FileFormat.V2.GreenshotFileV2`.

The new .greenshot file is just like the old ones a PNG file with binary data at the end. **The main difference is that now the rendered editor image is saved.** This includes all graphical elements in the editor. The raw screenshot from surface is also included in the binary data.

For copy and paste within the Greenshot Editor i used the same serialization mechanism as for the file format. 

More information can be found here: [Greenshot.Editor/FileFormat/readme.md](https://github.com/Christian-Schulz/greenshot/blob/feature/new_greenshot_file_format_v2/src/Greenshot.Editor/FileFormat/readme.md) .


### Backward compatibility

To support all old files there is the new namespace `Greenshot.Editor.FileFormat.V1.Legacy`. 
This is the only place where the BinaryFormatter is used.

Some rare `DrawableContainers` are serialized differently in Greenshot 1.2 and 1.3. I have implemented a workaround here that allows both variants to be read.

The main functionality, which can read and write every known file version, is implemented in the new `Greenshot.Editor.FileFormat.GreenshotFileVersionHandler`.


### Refactoring

I removed all [Serializable] attributes and most of serialization-specific code from `DrawableContainer` classes. The [Serializable] attributes are only needed in legacy-container classes.

All code from `Surface`, `ImageEditorForm` and `ImageIO` that dealt with reading and writing .greenhot files has been simplified and moved to `GreenshotFileFormatHandler`. (and to the new `GreenshotTemplateFormatHandler` for .gst files)

### Testing

I added a new test project.
The test data contains .greenshot files with all individually possible `DrawableContainers` for versions 1.2 (V1), 1.3 (V1), and V2.
The unit tests test the loading of each individual file.
There are unit test for `ConvertDomainToDto` and `ConvertDtoToDomain` for every `DrawableContainer`.
The `ConvertDomainToDto` test also checks container default values.
There are unit test that test the entire process from serialization to deserialization.
And there are a few unit tests to check the bugs I found along the way.


### Special notes

There is a `CursorContainer` that is not being used. I think this was intended for the mouse cursor, but for a capture with mouse cursor we use the `IconContainer`.
So, is the `CursorContainer` obsolete?

For the `MetafileContainer` i implemented some workarounds. For serialization with MessagePack we have to store the original source file, but GDI+ does not support saving .wmf or .emf files, because there is no encoder. So I had to create a replacement. During the deserialisation with BinaryFormatter of old file formats, i convert the internal metafile to a PNG. Visually in the editor, I didn't saw any difference. There is no problem with newly created `MetafileContainers`, as we have the original source file.

The BinaryFormatter is still used, but could easily be removed.

For the future, after switching to .NET9+, we could implement a new file format that is a valid PNG with the help of ImageSharp. That could contains the serialized data as PNG chunks. This would only require a few classes to be adapted.

---

I'm leaving the PR as a draft because I think there will be a few more adjustments based on the comments before it's finally finished.

_Please feel free to mark typos in my code._[](url)

Christian-Schulz committed:
New Greenshot file format

Replace Binaryformatter with MessagePack.
Add data transfer objects for all objects that are serialized.
Still support all previous file formats, first step to remove Binaryformatter.
Add new testproject with xunit.

Christian-Schulz committed:
Fix: Evaluate LoadDrawablesFromStream() before disposing the stream

Break after a container is found

Christian-Schulz committed:
extend SimpleServiceProvider, include the ability to remove services

Christian-Schulz committed:
Add lock for IniConfig.SectionMap

Christian-Schulz committed:
Merge branch 'bug/drag_drop_twice' into feature/new_greenshot_file_format_v2

Christian-Schulz committed:
Merge branch 'bug/lock_section_map' into feature/new_greenshot_file_format_v2

Christian-Schulz committed:
Merge branch 'extend_service_provider' into feature/new_greenshot_file_format_v2

Christian-Schulz committed:
Fix typo

jklingen commented:
Wow, that's impressive 🤩 I didn't have a closer look into the implementation yet, but the detailed description shows these PRs are carefully crafted. Thanks for your efforts 👍  

> First of all, I must apologize. 
lol, no - not really 😏 this is an enormous task and I totally understand that it cannot be sliced into smaller pieces than that.

I will try to have a look at it as soon as possible - given the large number of changes, this pull request shouldn't be left lying around for too long.

Christian-Schulz commented:
@FF-Brown, I had a merge conflict with your changes from #574, I'm pretty sure I re-implemented the same fix. Can you double check that?

FF-Brown commented:
@Christian-Schulz I don't see any changes to BinaryFormatterHelper in this PR, and that's the only file touched by #574. Where's the conflict?

Christian-Schulz commented:
> Where's the conflict?

I deleted Helpers/BinaryFormatterHelper.cs  and moved it to FileFormat/V1/Legacy/LegacySerializationBinder.cs

FF-Brown commented:
Ah didn't see that, will take a look

FF-Brown commented:
> I'm pretty sure I re-implemented the same fix

Yep looks good to me

Christian-Schulz commented:
> It is needed to change of version, go to 2.0 with this PR.

No, from the user's point of view, these changes are not that significant.

Christian-Schulz committed:
Fix typo

Christian-Schulz committed:
Merge branch 'release/1.3' into feature/new_greenshot_file_format_v2

Christian-Schulz committed:
Fix DuplicateSelectedElements in main and context menu

overwrites ObjectExtensions.Clone() and uses new serialisation

jklingen commented on src/Greenshot.Editor/Drawing/DrawableContainerList.cs:
```diff
@@ -780,5 +780,14 @@ public void AdjustToDpi(int dpi)
         }
+
+        /// <summary>
+        /// Creates a clone by using serialization 
```
```suggestion
        /// Creates a clone by using our own DTO/domain conversion logic.
```

Christian-Schulz commented on src/Greenshot.Editor/Drawing/ArrowContainer.cs:
```diff
@@ -56,7 +55,6 @@ public ArrowContainer(ISurface parent) : base(parent)
         protected override void InitializeFields()
         {
             AddField(GetType(), FieldType.LINE_THICKNESS, 2);
-            AddField(GetType(), FieldType.ARROWHEADS, 2);
```
yes, it was added twice.  Look 4 Lines after this. And field  `ARROWHEADS` is an enum.

Christian-Schulz committed:
Apply suggestions from code review

some typos

Co-authored-by: jklingen <jklingen@users.noreply.github.com>

Christian-Schulz commented on src/Greenshot.Base/IniFile/IniConfig.cs:
```diff
@@ -389,7 +397,12 @@ public static IEnumerable<string> IniSectionNames
+                List<string> sectionNames;
+                lock (SectionMapLock)
+                {
+                    sectionNames = [.. SectionMap.Keys];
```
Yes, I guess you have an old SDK-version installed (.NET 7 SDK or older). 
I use the current version .NET 9 SDK.

Lakritzator commented:
Ah yeah, using a custom chunk in a .png was on my plan years ago, you can still see it here: https://github.com/SixLabors/ImageSharp/issues/251
You can see I linked it with a Greenshot issue, which I created to discuss the next format: https://github.com/greenshot/greenshot/issues/375

But the ImageSharp change was never finished, the issue is still open :/

The .gsz is just an idea, .greenshotzip...
It's like "comic book .zip" -> .CBZ

My goal was to keep the original images which were dragged etc onto the canvas, so we can edit them however it would be possible. So also SVG... With quality loss.

jklingen commented:
> What I really liked about the old file format was that it was actually a PNG file. It could be opened with any image viewer or browser.

Fair point. I had that completely off my radar. (Personally I mostly create one-time screenshots, so I never use .greenshot files.). As developer *and* user, I generelly prefer transparent file formats, but I can imagine that the PNG part is an important feature for quite some users, esp. when using Greenshot for documentation purposes.

Christian-Schulz committed:
Merge remote-tracking branch 'Greenshot/main' into feature/new_greenshot_file_format_v2

# Conflicts:
#	src/Greenshot.Editor/Drawing/DrawableContainerList.cs

jklingen commented:
> only the screenshot can be seen, all elements like lines, text, and effects are not visible on it.

Good point. The value of the preview is limited if it doesnt show what the file content looks like.

> I would now recommend the ZIP/JSON format.
> 
> That would be my first thought about the file structure in zip format.

Sounds like we're all team zipped JSON by now 😁 for me, the suggested file structure looks fine - and I think it's pretty much like what @Lakritzator had in mind?

Christian-Schulz committed:
Merge branch 'main' into feature/new_greenshot_file_format_v2

# Conflicts:
#	src/Greenshot.Editor/Drawing/Surface.cs
#	src/Greenshot.Editor/Greenshot.Editor.csproj
#	src/Greenshot.Editor/Helpers/BinaryFormatterHelper.cs
#	src/Greenshot.sln

Christian-Schulz committed:
Support the new EmojiContainer

Christian-Schulz committed:
Add support for EmojiContainer and greenshot 1.4 files

Add backwards compatibility for new fields in StepLabelContainer

Christian-Schulz committed:
Minimal running JSON/ZIP implementation

Christian-Schulz committed:
Save images in zip as file,

serialize color as hex code, reorder properties and rename types discriminator for better readability in JSON

Christian-Schulz commented:
FYI: This is still Work-In-Process.

The current version can save and load JSON/ZIP files. If you want, you can try it.

<details><summary>Sample content.json </summary>

```json
{
  "formatVersion": "V3",
  "schemaVersion": 1,
  "imagePath": "Images/image_1.png",
  "renderedImagePath": "Images/image_2.png",
  "containerList": {
    "containerList": [
      {
        "$type": "Line",
        "left": 240,
        "top": 175,
        "width": 37,
        "height": 40,
        "fields": [
          {
            "fieldTypeName": "LINE_THICKNESS",
            "value": {
              "$type": "Int",
              "value": 4
            },
            "scope": "LineContainer"
          },
          {
            "fieldTypeName": "LINE_COLOR",
            "value": {
              "$type": "Color",
              "color": "#FF0000FF"
            },
            "scope": "LineContainer"
          },
          {
            "fieldTypeName": "SHADOW",
            "value": {
              "$type": "Bool",
              "value": false
            },
            "scope": "LineContainer"
          }
        ]
      },
      {
        "$type": "Speechbubble",
        "text": null,
        "storedTargetGripperLocation": {
          "x": 45,
          "y": 312
        },
        "left": 58,
        "top": 281,
        "width": 75,
        "height": -30,
        "fields": [
          {
            "fieldTypeName": "LINE_THICKNESS",
            "value": {
              "$type": "Int",
              "value": 3
            },
            "scope": "SpeechbubbleContainer"
          },
          {
            "fieldTypeName": "LINE_COLOR",
            "value": {
              "$type": "Color",
              "color": "#FF0000FF"
            },
            "scope": "SpeechbubbleContainer"
          },
          {
            "fieldTypeName": "SHADOW",
            "value": {
              "$type": "Bool",
              "value": true
            },
            "scope": "SpeechbubbleContainer"
          },
          {
            "fieldTypeName": "FONT_ITALIC",
            "value": {
              "$type": "Bool",
              "value": false
            },
            "scope": "SpeechbubbleContainer"
          },
          {
            "fieldTypeName": "FONT_BOLD",
            "value": {
              "$type": "Bool",
              "value": false
            },
            "scope": "SpeechbubbleContainer"
          },
          {
            "fieldTypeName": "FILL_COLOR",
            "value": {
              "$type": "Color",
              "color": "#FFFFFFFF"
            },
            "scope": "SpeechbubbleContainer"
          },
          {
            "fieldTypeName": "FONT_FAMILY",
            "value": {
              "$type": "String",
              "value": "Arial"
            },
            "scope": "SpeechbubbleContainer"
          },
          {
            "fieldTypeName": "FONT_SIZE",
            "value": {
              "$type": "Single",
              "value": 20
            },
            "scope": "SpeechbubbleContainer"
          },
          {
            "fieldTypeName": "TEXT_HORIZONTAL_ALIGNMENT",
            "value": {
              "$type": "StringAlignment",
              "value": "Center"
            },
            "scope": "SpeechbubbleContainer"
          },
          {
            "fieldTypeName": "TEXT_VERTICAL_ALIGNMENT",
            "value": {
              "$type": "StringAlignment",
              "value": "Center"
            },
            "scope": "SpeechbubbleContainer"
          }
        ]
      },
      {
        "$type": "StepLbel",
        "number": 1,
        "counterStart": 1,
        "left": 164,
        "top": 248,
        "width": 30,
        "height": 30,
        "fields": [

```
</details>

Christian-Schulz committed:
dedicated file for meta information

Christian-Schulz committed:
generic way to save all images in zip

use attributes in dto for that

Christian-Schulz committed:
Remove MessagePack

temporarily disabled serialization V2

Christian-Schulz committed:
V2 is now JSON/ZIP, reactivate template support

Christian-Schulz committed:
ability to save .gsa,  reactivate serialization tests , create JSON/ZIP test files

Christian-Schulz committed:
reactivate template tests, documentation for all V2 methodes,

a bit refactoring and some minor loading fixes

Christian-Schulz committed:
reactivate copy&paste

Christian-Schulz committed:
Doc

Christian-Schulz committed:
Rewrite FileFormat readme for JSON/ZIP

Christian-Schulz committed:
Refactor and fix extension handling for image properties

Christian-Schulz committed:
License header outdated

Christian-Schulz committed:
Fix test files for MetafileContainer; simple exclude of old .greenshot files in SaveDialog

Christian-Schulz committed:
Merge remote-tracking branch 'origin_Chris/main' into feature/new_greenshot_file_format_v2

# Conflicts:
#	src/Greenshot.Base/Controls/SaveImageFileDialog.cs
#	src/Greenshot.Base/Core/Enums/OutputFormat.cs
#	src/Greenshot.Editor/Drawing/CursorContainer.cs
#	src/Greenshot.Editor/Drawing/Fields/AbstractFieldHolder.cs
#	src/Greenshot.Editor/Drawing/StepLabelContainer.cs
#	src/Greenshot.Editor/Drawing/Surface.cs
#	src/Greenshot.Editor/EditorInitialize.cs
#	src/Greenshot.Editor/FileFormatHandlers/GreenshotFileFormatHandler.cs
#	src/Greenshot.Editor/Helpers/BinaryFormatterHelper.cs
#	src/Greenshot.sln

Christian-Schulz committed:
Add support for new CursorContainer

serialization, dto and test cases

Christian-Schulz committed:
Fix Call V2 methode in V2 test directly

Christian-Schulz commented:
I would say the PR is ready for the next review. 

Greenshot now serializes all data in JSON and saves it in a zip archive (`.gsa` file) .
The old `.greenshot` files can still be read but can no longer be saved.

Greenshot uses the same serialization mechanism for the clipboard. This uses JSON without a ZIP file.

General information about the new format can be found here: [Greenshot.Editor/FileFormat/readme.md](https://github.com/Christian-Schulz/greenshot/blob/feature/new_greenshot_file_format_v2/src/Greenshot.Editor/FileFormat/readme.md) .

Feel free to ask me questions.

Christian-Schulz committed:
Merge branch 'main' into feature/new_greenshot_file_format_v2

# Conflicts:
#	src/Greenshot.Editor/Drawing/StepLabelContainer.cs

Christian-Schulz committed:
Merge branch 'main' into feature/new_greenshot_file_format_v2

# Conflicts:
#	src/Greenshot.Editor/FileFormatHandlers/GreenshotFileFormatHandler.cs

Christian-Schulz committed:
Using RecyclableMemoryStreamFactory in Greenshot file handler classes

Christian-Schulz committed:
Merge branch 'main' into feature/new_greenshot_file_format_v2

# Conflicts:
#	src/Greenshot.Editor/Drawing/ArrowContainer.cs
#	src/Greenshot.Editor/Drawing/CursorContainer.cs
#	src/Greenshot.Editor/Drawing/DrawableContainer.cs
#	src/Greenshot.Editor/Drawing/EllipseContainer.cs
#	src/Greenshot.Editor/Drawing/Emoji/EmojiContainer.cs
#	src/Greenshot.Editor/Drawing/Fields/AbstractFieldHolder.cs
#	src/Greenshot.Editor/Drawing/ImageContainer.cs
#	src/Greenshot.Editor/Drawing/LineContainer.cs
#	src/Greenshot.Editor/Drawing/RectangleContainer.cs
#	src/Greenshot.Editor/Drawing/SpeechbubbleContainer.cs
#	src/Greenshot.Editor/Drawing/Surface.cs
#	src/Greenshot.Editor/Drawing/TextContainer.cs
#	src/Greenshot.Editor/Helpers/BinaryFormatterHelper.cs
#	src/Greenshot.sln
